### ISSUE TITLE
Bundler 1.13.4 and 1.13.5 private method `gem'

### ISSUE THREAD

**Issue by schneems** [bundler/bundler#5102]
In bundler prior to 1.13 there was a public `gem` method in `gems/bundler-1.11.2/lib/bundler/rubygems_integration.rb:283`

In Bundler 1.13.5 it's private, which breaks existing applications using that declaration. 

For example that was used in Rails apps in the config section

```ruby
module Foo
  class Application < Rails::Application
    config.gem 'rake'
  end
end
```

We need to make this public again. If it needs to be private we should go through a deprecation cycle.

**Comment by schneems** [comment]
Looks like it was modified in https://github.com/bundler/bundler/commit/c144eb67490d11db85a808661402a54ee297ba0c#diff-f271319d553ee0af23e993e254a92c6cR305

**Comment by segiddins** [comment]
```ruby
#!/usr/bin/env ruby
require 'rubygems'
class C
  def f
    self.gem 'foo'
  end
end

C.new.f
```

It's private without even requiring bundler

**Comment by schneems** [comment]
This was with Ruby 1.9.3

```
$ gem -v
1.8.23.2
$ ruby -v
ruby 1.9.3p545 (2014-02-24) [x86_64-darwin14.5.0]
```

Bundler 1.11.2 (works): https://gist.github.com/schneems/97c4b525a66bc913ce51377dcafc5fd9
Bundler 1.13.5 (fails): https://gist.github.com/schneems/7e14ae8299c5940b6f6b0fe303163292

**Comment by schneems** [comment]
Also fails for Ruby 2.3.1. Here's an example app:

https://github.com/schneems/bundler-private-gem-method-issue

**Comment by segiddins** [comment]
https://github.com/rubygems/rubygems/blob/master/lib/rubygems/core_ext/kernel_gem.rb#L72

**Comment by schneems** [comment]
Bundler made it not private. It works in older versions of bundler because it didn't care about method visibility:

```
::Kernel.send(:define_method, :gem) do |dep, *reqs|
```

**Comment by segiddins** [comment]
That was a bug in the old versions of bundler.

**Comment by schneems** [comment]
I agree, it's also a bug that a good number of people were relying on without knowing about. Can we un-private that method and deprecate it?

**Comment by segiddins** [comment]
We can't deprecate it, it's a RubyGems method!

**Comment by indirect** [comment]
@segiddins you are absolutely, 100% correct. Unfortunately, it's more important that apps keep working than to do the 100% correct thing. The explicit goal of the Bundler project is to help people manage their dependencies _predictably_ and _reliably_. Predictability and reliability is only possible when we are willing to undo bugfixes that break apps until we can deprecate them.

@schneems thanks for pointing out that this breaks existing apps. We'll make the method public again in a bugfix release for 1.13.x.

**Comment by schneems** [comment]
Awesome, thanks!

![](https://media.giphy.com/media/k9qu4GDdGVDLa/giphy.gif)

**Comment by segiddins** [comment]
Released in 1.13.6

**Comment by schneems** [comment]
Thanks! I'll get this on Heroku master ASAP
